cli,sql: do not exit interactive shells with Ctrl+C

knz opened the pull request:
First commit from #76434.
Needed for #76433.

This change is a preface to making the interactive shell support
Ctrl+C to cancel a currently executing query.

When that happens, we need to ensure that Ctrl+C *only* cancels the
current query, and not the entire shell. This is because there is no
deterministic boundary in time for a user to decide whether the key
press will cancel only the query, or stop the shell. If they care
about keeping their shell alive, they would never "dare" to press
Ctrl+C while a query is executing.

Incidentally, this is also what `psql` does, for pretty much the same
reason.

Note that the new behavior is restricted to *interactive* shells, when
stdin is a terminal. The behavior for non-interactive shells remains
unchanged, where Ctrl+C will terminate everything. This is also what
other SQL shells do.

Release note (cli change): `cockroach sql` (and `demo`) now continue
to accept user input when Ctrl+C is pressed at the interactive prompt
and the current input line is empty. Previously, it would terminate
the shell.  To terminate the shell, the client-side command `\q` is
still supported. The user can also terminate the input altogether via
EOF (Ctrl+D).
The behavior in non-interactive use remains unchanged.

knz committed:
cli,sql: do not exit interactive shells with Ctrl+C

This change is a preface to making the interactive shell support
Ctrl+C to cancel a currently executing query.

When that happens, we need to ensure that Ctrl+C *only* cancels the
current query, and not the entire shell. This is because there is no
deterministic boundary in time for a user to decide whether the key
press will cancel only the query, or stop the shell. If they care
about keeping their shell alive, they would never "dare" to press
Ctrl+C while a query is executing.

Incidentally, this is also what `psql` does, for pretty much the same
reason.

Note that the new behavior is restricted to *interactive* shells, when
stdin is a terminal. The behavior for non-interactive shells remains
unchanged, where Ctrl+C will terminate everything. This is also what
other SQL shells do.

Release note (cli change): `cockroach sql` (and `demo`) now continue
to accept user input when Ctrl+C is pressed at the interactive prompt
and the current input line is empty. Previously, it would terminate
the shell.  To terminate the shell, the client-side command `\q` is
still supported. The user can also terminate the input altogether via
EOF (Ctrl+D).
The behavior in non-interactive use remains unchanged.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/76427)
<!-- Reviewable:end -->

knz committed:
cli: modernize the SQL API calls

This brings the cli (sub-)packages kicking and screaming into the
post-go1.8 era, by adding the missing context arguments.

Release note (bug fix): Some of the `cockroach node` subcommands did
not handle `--timeout` properly. This is now fixed.

rafiss reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76427)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz, @otan, and @shermanCRL)

___
*[pkg/cli/clisqlshell/sql.go, line 1025 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76427#-MvdsZTnECYN9jGOCBqq:-MvdsZTnECYN9jGOCBqr:biwvntw) ([raw file](https://github.com/cockroachdb/cockroach/blob/7064116c335b9de62ca3cc3a790d02d339a000f4/pkg/cli/clisqlshell/sql.go#L1025)):*
> ```Go
> 
> 		// If a human is looking, tell them that quitting is done in another way.
> 		if c.sqlExecCtx.TerminalOutput {
> ```

change lgtm, but just for my own education, when could `IsInteractive` be true, but `TerminalOutput` be false?


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76427)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan, @rafiss, and @shermanCRL)

___
*[pkg/cli/clisqlshell/sql.go, line 1025 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76427#-MvdsZTnECYN9jGOCBqq:-Mvdsoal91H4UY0g2_RT:bqb9gq1) ([raw file](https://github.com/cockroachdb/cockroach/blob/7064116c335b9de62ca3cc3a790d02d339a000f4/pkg/cli/clisqlshell/sql.go#L1025)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

change lgtm, but just for my own education, when could `IsInteractive` be true, but `TerminalOutput` be false?
</blockquote></details>

`cockroach sql >results.out`
then type `select 123;`, enter, `\q`, enter

that's interactive, but no terminal output.


<!-- Sent from Reviewable.io -->

knz commented:
bors r=rafiss
